cmd/mutate: ability to remove lbls and anntns

Dentrax opened the pull request:
Fixes #1062

Signed-off-by: Furkan <furkan.turkal@trendyol.com>

Dentrax commented:
@googlebot I signed it!

imjasonh commented on cmd/crane/cmd/mutate.go:
```diff
@@ -65,16 +66,20 @@ func NewCmdMutate(options *[]crane.Option) *cobra.Command {
 
 			for k, v := range labels {
-				cfg.Config.Labels[k] = v
+				if !removeMode {
```
I think I'd prefer to phrase this as:

```go
if removeMode {
  delete(cfg.Config.Labels, k)
} else {
  cfg.Config.Labels[k] = v
}
```

imjasonh commented on cmd/crane/cmd/mutate.go:
```diff
@@ -91,7 +96,17 @@ func NewCmdMutate(options *[]crane.Option) *cobra.Command {
+					delete(manifest.Annotations, k)
+				}
+			} else {
+				img = mutate.Annotations(img, annotations)
```
Should we have some `mutate.RemoveAnnotations` method that we use here? I think it's confusing and inconsistent at least, and a bug at most, that `img.Manifest()` returns a mutable `manifest` that modifies the `img`. When we eventually fix that behavior, we'll want to make sure we don't break this code.

imjasonh commented on cmd/crane/cmd/mutate.go:
```diff
@@ -119,20 +134,28 @@ func NewCmdMutate(options *[]crane.Option) *cobra.Command {
 // splitKeyVals splits key value pairs which is in form hello=world
-func splitKeyVals(kvPairs []string) (map[string]string, error) {
+// if requireSplit set true, only gathers keys otherwise.
+func splitKeyVals(kvPairs []string, enforceSplit bool) (map[string]string, error) {
```
Instead of repurposing this method, and making it do two things depending on `enforceSplit`, maybe let's just have a separate method that translates a `[]string` -> `map[string]string` (or `map[string]struct{}`, even).

imjasonh reviewed (commented):
Most of this feedback is a bit less relevant in light of Jon's advice to support `--unlabel` and `--unannotate`, but some of it might still be useful after incorporating that feedback.

Dentrax commented:
Hey @imjasonh! I couldn't find a time to get to it due to our national holiday. I updated the code according to your reviews! I also added `RemoveAnnotations()` function into `mutate` package. :)

Commands now should be work as expected:

1. Mutate the image:
```bash
$ go run . mutate --label foo=bar --annotation baz=qux furkanturkal/busybox:0.2.0`
```

2. Check the mutations:
```bash
$ crane manifest furkanturkal/busybox:0.2.0 | jq
...
"annotations": {
  "baz": "qux"
}

$ crane config furkanturkal/busybox:0.2.0 | jq
...
"Labels": {
  "foo": "bar"
}
```

3. Remote both _label_s and _annotation_s:
```bash
$ go run . mutate --unlabel foo --unannotate baz furkanturkal/busybox:0.2.0
# check again if both removed
```

**P.S:** Minor flag name inconsistency has occurred: `--unannotate` and `--annotation`. Should we re-name one of them? (`tation` and `tate`)

---

In the `RemoveAnnotations` function, I had to traverse _annotations_ because I could not figure it out how to make it work using `match.Matcher` instead of _for-loop_. I was try to add a new field into `image` struct, as `index` struct did. And in the `compute()` function, I was trying to figure it out using this `matcher` function to remove (filter) annotations. I tried to call this new function by passing `img, match.Annotation("foo", "bar")` params.

```go
	if i.remove != nil {
		var cleanedManifests []v1.Descriptor
		for _, m := range manifests {
			if !i.remove(m) {
				cleanedManifests = append(cleanedManifests, m)
			}
		}
		manifests = cleanedManifests
	}
```

But somehow, `img.Manifest().Annotations` returned `nil` (which caused by `if !i.remove(m)` never returned _true_) ; but still, i was able to see the `foo: bar` value in the manifest: `$ crane manifest furkanturkal/busybox:0.2.0 | jq`. Maybe, I missed some important point. 🤷‍♂️

jonjohnsonjr commented on pkg/v1/mutate/mutate.go:
```diff
@@ -101,6 +101,23 @@ func RemoveManifests(base v1.ImageIndex, matcher match.Matcher) v1.ImageIndex {
 }
 
+// RemoveAnnotations removes annotations the provided v1.Image
+func RemoveAnnotations(base v1.Image, annotations []string) (v1.Image, error) {
```
Don't return an error here, we can do this lazily. We'll need to store this list of annotations to delete in image{} (maybe call it `unannotations` and call `delete()` around here: https://github.com/google/go-containerregistry/blob/eca1cd86affe459c22a2c7a818769ea70d13496e/pkg/v1/mutate/image.go#L157

I also wonder if we want to give this the same treatment as https://github.com/google/go-containerregistry/pull/1082

Dentrax replied:
That makes sense! Now, it's much cleaner. 👍
